### PR TITLE
Transition activation_functions_test.cpp to Catch

### PR DESCRIPTION
**Pull request by rcurtin**
This was a quick attempt to help with #2523.  I did this partially by hand, partially with regular expressions.  I believe that it would be possible to do this all with regular expressions (plus a little bit of checking by hand afterwards). :+1:

### CODE REVIEW

File: src/mlpack/tests/activation_functions_test.cpp [hunk 1]
```diff
@@ -59,16 +57,16 @@ void CheckActivationCorrect(const arma::colvec input,
   // Test the activation function using a single value as input.
   for (size_t i = 0; i < target.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(ActivationFunction::Fn(input.at(i)),
-        target.at(i), 1e-3);
+    REQUIRE(ActivationFunction::Fn(input.at(i)) ==
+        Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:61]:
```suggestion
    REQUIRE(ActivationFunction::Fn(input.at(i)) ==
        Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 2]
```diff
@@ -59,16 +57,16 @@ void CheckActivationCorrect(const arma::colvec input,
...
   }
 
   // Test the activation function using the entire vector as input.
   arma::colvec activations;
   ActivationFunction::Fn(input, activations);
   for (size_t i = 0; i < activations.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(activations.at(i), target.at(i), 1e-3);
+    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:69]:
```suggestion
    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 3]
```diff
@@ -87,16 +85,16 @@ void CheckDerivativeCorrect(const arma::colvec input,
   // Test the calculation of the derivatives using a single value as input.
   for (size_t i = 0; i < target.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(ActivationFunction::Deriv(input.at(i)),
-        target.at(i), 1e-3);
+    REQUIRE(ActivationFunction::Deriv(input.at(i)) ==
+        Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:89]:
```suggestion
    REQUIRE(ActivationFunction::Deriv(input.at(i)) ==
        Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 4]
```diff
@@ -87,16 +85,16 @@ void CheckDerivativeCorrect(const arma::colvec input,
...
   }
 
   // Test the calculation of the derivatives using the entire vector as input.
   arma::colvec derivatives;
   ActivationFunction::Deriv(input, derivatives);
   for (size_t i = 0; i < derivatives.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(derivatives.at(i), target.at(i), 1e-3);
+    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:97]:
```suggestion
    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 5]
```diff
@@ -114,8 +112,8 @@ void CheckInverseCorrect(const arma::colvec input)
     // Test the calculation of the inverse using a single value as input.
   for (size_t i = 0; i < input.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(ActivationFunction::Inv(ActivationFunction::Fn(
-        input.at(i))), input.at(i), 1e-3);
+    REQUIRE(ActivationFunction::Inv(ActivationFunction::Fn(input.at(i))) ==
+        Approx(input.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:116]:
```suggestion
    REQUIRE(ActivationFunction::Inv(ActivationFunction::Fn(input.at(i))) ==
        Approx(input.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 6]
```diff
@@ -125,7 +123,7 @@ void CheckInverseCorrect(const arma::colvec input)
 
   for (size_t i = 0; i < input.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(activations.at(i), input.at(i), 1e-3);
+    REQUIRE(activations.at(i) == Approx(input.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:126]:
```suggestion
    REQUIRE(activations.at(i) == Approx(input.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 7]
```diff
@@ -146,7 +144,7 @@ void CheckHardTanHActivationCorrect(const arma::colvec input,
   htf.Forward(input, activations);
   for (size_t i = 0; i < activations.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(activations.at(i), target.at(i), 1e-3);
+    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:147]:
```suggestion
    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 8]
```diff
@@ -172,7 +170,7 @@ void CheckHardTanHDerivativeCorrect(const arma::colvec input,
 
   for (size_t i = 0; i < derivatives.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(derivatives.at(i), target.at(i), 1e-3);
+    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:173]:
```suggestion
    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 9]
```diff
@@ -194,7 +192,7 @@ void CheckLeakyReLUActivationCorrect(const arma::colvec input,
   lrf.Forward(input, activations);
   for (size_t i = 0; i < activations.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(activations.at(i), target.at(i), 1e-3);
+    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:195]:
```suggestion
    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 10]
```diff
@@ -220,7 +218,7 @@ void CheckLeakyReLUDerivativeCorrect(const arma::colvec input,
   lrf.Backward(input, error, derivatives);
   for (size_t i = 0; i < derivatives.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(derivatives.at(i), target.at(i), 1e-3);
+    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:221]:
```suggestion
    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 11]
```diff
@@ -242,7 +240,7 @@ void CheckELUActivationCorrect(const arma::colvec input,
   lrf.Forward(input, activations);
   for (size_t i = 0; i < activations.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(activations.at(i), target.at(i), 1e-3);
+    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:243]:
```suggestion
    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 12]
```diff
@@ -268,7 +266,7 @@ void CheckELUDerivativeCorrect(const arma::colvec input,
   lrf.Backward(activations, error, derivatives);
   for (size_t i = 0; i < derivatives.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(derivatives.at(i), target.at(i), 1e-3);
+    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:269]:
```suggestion
    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 13]
```diff
@@ -290,7 +288,7 @@ void CheckPReLUActivationCorrect(const arma::colvec input,
   prelu.Forward(input, activations);
   for (size_t i = 0; i < activations.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(activations.at(i), target.at(i), 1e-3);
+    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:291]:
```suggestion
    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 14]
```diff
@@ -316,7 +314,7 @@ void CheckPReLUDerivativeCorrect(const arma::colvec input,
   prelu.Backward(input, error, derivatives);
   for (size_t i = 0; i < derivatives.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(derivatives.at(i), target.at(i), 1e-3);
+    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:317]:
```suggestion
    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 15]
```diff
@@ -340,9 +338,9 @@ void CheckPReLUGradientCorrect(const arma::colvec input,
   // This error vector will be set to 1 to get the gradient.
   arma::colvec error = arma::ones<arma::colvec>(input.n_elem);
   prelu.Gradient(input, error, gradient);
-  BOOST_REQUIRE_EQUAL(gradient.n_rows, 1);
-  BOOST_REQUIRE_EQUAL(gradient.n_cols, 1);
-  BOOST_REQUIRE_CLOSE(gradient(0), target(0), 1e-3);
+  REQUIRE(gradient.n_rows == 1);
+  REQUIRE(gradient.n_cols == 1);
+  REQUIRE(gradient(0) == Approx(target(0)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:343]:
```suggestion
  REQUIRE(gradient(0) == Approx(target(0)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 16]
```diff
@@ -362,7 +360,7 @@ void CheckHardShrinkActivationCorrect(const arma::colvec input,
   hardshrink.Forward(input, activations);
   for (size_t i = 0; i < activations.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(activations.at(i), target.at(i), 1e-3);
+    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:363]:
```suggestion
    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 17]
```diff
@@ -388,7 +386,7 @@ void CheckHardShrinkDerivativeCorrect(const arma::colvec input,
   hardshrink.Backward(input, error, derivatives);
   for (size_t i = 0; i < derivatives.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(derivatives.at(i), target.at(i), 1e-3);
+    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:389]:
```suggestion
    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 18]
```diff
@@ -410,7 +408,7 @@ void CheckSoftShrinkActivationCorrect(const arma::colvec input,
   softshrink.Forward(input, activations);
   for (size_t i = 0; i < activations.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(activations.at(i), target.at(i), 1e-3);
+    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:411]:
```suggestion
    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 19]
```diff
@@ -436,15 +434,15 @@ void CheckSoftShrinkDerivativeCorrect(const arma::colvec input,
   softshrink.Backward(input, error, derivatives);
   for (size_t i = 0; i < derivatives.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(derivatives.at(i), target.at(i), 1e-3);
+    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:437]:
```suggestion
    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 20]
```diff
@@ -531,7 +529,7 @@ void CheckCELUActivationCorrect(const arma::colvec input,
   lrf.Forward(input, activations);
   for (size_t i = 0; i < activations.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(activations.at(i), target.at(i), 1e-3);
+    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:532]:
```suggestion
    REQUIRE(activations.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 21]
```diff
@@ -557,14 +555,14 @@ void CheckCELUDerivativeCorrect(const arma::colvec input,
   lrf.Backward(activations, error, derivatives);
   for (size_t i = 0; i < derivatives.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(derivatives.at(i), target.at(i), 1e-3);
+    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:558]:
```suggestion
    REQUIRE(derivatives.at(i) == Approx(target.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 22]
```diff
@@ -744,18 +742,20 @@ BOOST_AUTO_TEST_CASE(CReLUFunctionTest)
   crelu.Backward(desiredActivations, error, derivatives);
   for (size_t i = 0; i < activations.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(activations.at(i), desiredActivations.at(i), 1e-3);
+    REQUIRE(activations.at(i) ==
+        Approx(desiredActivations.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:746]:
```suggestion
    REQUIRE(activations.at(i) ==
        Approx(desiredActivations.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 23]
```diff
@@ -744,18 +742,20 @@ BOOST_AUTO_TEST_CASE(CReLUFunctionTest)
...
   }
   for (size_t i = 0; i < derivatives.n_elem; ++i)
   {
-    BOOST_REQUIRE_CLOSE(derivatives.at(i), desiredDerivatives.at(i), 1e-3);
+    REQUIRE(derivatives.at(i) ==
+        Approx(desiredDerivatives.at(i)).epsilon(1e-3));
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:751]:
```suggestion
    REQUIRE(derivatives.at(i) ==
        Approx(desiredDerivatives.at(i)).epsilon(1e-3 / 100));
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 24]
```diff
@@ -33,14 +33,12 @@
 #include <mlpack/methods/ann/activation_functions/poisson1_function.hpp>
 #include <mlpack/methods/ann/activation_functions/gaussian_function.hpp>
 
-#include <boost/test/unit_test.hpp>
-#include "test_tools.hpp"
```
Comment (jeffin143) [inline src/mlpack/tests/activation_functions_test.cpp:37]: shouldn't we add `#include "test_catch_tools.hpp"` here ?

File: src/mlpack/tests/activation_functions_test.cpp (hunk 23 above)
Comment (birm) [reply src/mlpack/tests/activation_functions_test.cpp:751]: Wouldn't you just use 1e-6 in that case?

Comment (jeffin143) [reply src/mlpack/tests/activation_functions_test.cpp:751]: Yes that would work too

Comment (birm) [reply src/mlpack/tests/activation_functions_test.cpp:751]: Yes, my bad. 1e-5 :laughing:

File: src/mlpack/tests/activation_functions_test.cpp (hunk 24 above)
Comment (rcurtin) [reply src/mlpack/tests/activation_functions_test.cpp:37]: You're right, thanks!  0d651e29

File: src/mlpack/tests/activation_functions_test.cpp (hunk 1 above)
Comment (rcurtin) [reply src/mlpack/tests/activation_functions_test.cpp]: Ahhh, I misunderstood the catch tolerance.  Thanks for catching this (no pun intended :smile:).  I've updated all the tolerances to divide by 100 in dfdd9e764.

File: src/mlpack/tests/activation_functions_test.cpp [hunk 25]
```diff
@@ -33,14 +33,12 @@
...
+#include "catch.hpp"
+#include "test_catch_tools.hpp"
```
Comment (zoq) [inline src/mlpack/tests/activation_functions_test.cpp:37]: Looks like we can remove the header file, because we don't use `CheckMatrices`.

Comment (jeffin143) [reply src/mlpack/tests/activation_functions_test.cpp:37]: Agreed

Comment (rcurtin) [reply src/mlpack/tests/activation_functions_test.cpp]: Done in 2bb130e.

**Comment by rcurtin** [comment]
Unlucky (unrelated) test failure, so I'll go ahead and merge anyway. :+1:

**Pull request merged**